File: app/src/main/java/arcus/app/device/settings/fragment/HaloRoomFragment.java
/*
 *  Copyright 2019 Arcus Project.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package arcus.app.device.settings.fragment;

import android.app.Activity;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import androidx.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;

import arcus.app.common.fragments.BaseFragment;
import arcus.cornea.CorneaClientFactory;
import arcus.cornea.device.smokeandco.HaloController;
import arcus.cornea.provider.DeviceModelProvider;
import com.iris.client.model.DeviceModel;
import arcus.app.R;
import arcus.app.common.adapters.CheckableListAdapter;
import arcus.app.common.backstack.BackstackManager;
import arcus.app.common.popups.AlertPopup;
import arcus.app.common.view.Version1Button;
import arcus.app.common.view.Version1ButtonColor;
import arcus.app.common.view.Version1TextView;

import java.util.List;

public class HaloRoomFragment extends BaseFragment implements HaloController.Callback, CheckableListAdapter.AdapterCallback {

    private static final String DEVICE_ADDRESS = "DEVICE_ADDRESS";
    private final static String EDIT_MODE = "EDIT_MODE";

    private ListView list;
    private Version1TextView listTitle;
    private Version1TextView listSubTitle;
    private Version1Button nextButton;
    HaloController haloController;
    HaloRoomFragment frag;
    CheckableListAdapter adapter;
    private String[] roomTypes;
    private String selection;
    private List<String> rooms;

    @NonNull
    public static HaloRoomFragment newInstance (String deviceAddress, boolean isEditMode) {
        HaloRoomFragment fragment = new HaloRoomFragment();

        Bundle bundle = new Bundle();
        bundle.putString(DEVICE_ADDRESS, deviceAddress);
        bundle.putBoolean(EDIT_MODE, isEditMode);
        fragment.setArguments(bundle);

        return fragment;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);

        list = (ListView) view.findViewById(R.id.list);
        nextButton = (Version1Button) view.findViewById(R.id.next_button);
        nextButton.setVisibility(isEditMode() ? View.GONE : View.VISIBLE);

        listTitle = (Version1TextView) view.findViewById(R.id.list_title);
        listTitle.setText(getString(R.string.halo_room_title));

        listSubTitle = (Version1TextView) view.findViewById(R.id.list_subtitle);
        listSubTitle.setText(getString(R.string.halo_room_subtitle));
        listSubTitle.setVisibility(View.VISIBLE);

        return view;
    }

    @Override
    public void onResume () {
        super.onResume();

        frag = this;

        nextButton.setColorScheme(isEditMode() ? Version1ButtonColor.WHITE : Version1ButtonColor.BLACK);
        listTitle.setTextColor(isEditMode() ? Color.WHITE : Color.BLACK);
        listSubTitle.setTextColor(isEditMode() ? Color.WHITE : Color.BLACK);
        nextButton.setText(isEditMode() ? getString(R.string.save_text) : getString(R.string.pairing_next));
        nextButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(adapter.getSelectedItem() == -1) {
                    AlertPopup alertPopup = AlertPopup.newInstance(
                            getString(R.string.halo_post_pairing_room_error_title),
                            getString(R.string.halo_post_pairing_room_error_body),
                            null,
                            null,
                            new AlertPopup.AlertButtonCallback() {
                                @Override public boolean topAlertButtonClicked() { return false; }
                                @Override public boolean bottomAlertButtonClicked() { return false; }
                                @Override public boolean errorButtonClicked() { return false; }

                                @Override public void close() {
                                    BackstackManager.getInstance().navigateBack();
                                }
                            }
                    );
                    alertPopup.setCloseButtonVisible(true);
                    BackstackManager.getInstance().navigateToFloatingFragment(alertPopup, alertPopup.getClass().getCanonicalName(), true);
                    return;
                }
            }
        });

        String deviceAddress = getArguments().getString(DEVICE_ADDRESS);

        haloController = new HaloController(
                DeviceModelProvider.instance().getModel(deviceAddress == null ? "DRIV:dev:" : deviceAddress),
                CorneaClientFactory.getClient(),
                null
        );
        haloController.setCallback(this);

        Activity activity = getActivity();
        if (activity != null) {
            activity.setTitle(getTitle());
        }

    }

    @NonNull
    @Override
    public String getTitle() {
        return getString(R.string.halo_post_pairing_info_title);
    }

    @Override
    public Integer getLayoutId() {
        return R.layout.fragment_checkable_list;
    }

    private boolean isEditMode() {
        return getArguments().getBoolean(EDIT_MODE);
    }

    @Override
    public void onError(Throwable throwable) {

    }

    @Override
    public void onSuccess(DeviceModel deviceModel) {
        rooms = haloController.getRoomTypes();

        roomTypes = rooms.toArray(new String[rooms.size()]);

        for(int nRoom = 0; nRoom < roomTypes.length; nRoom++) {
            roomTypes[nRoom] = roomTypes[nRoom].toUpperCase();
        }

        int defaultSelection = -1;
        if(isEditMode()) {
            selection = haloController.getSelectedRoomType();
        }
        if(selection != null) {
            defaultSelection = rooms.indexOf(selection);
        }
        adapter = new CheckableListAdapter(getActivity(), R.layout.cell_checkable_item, roomTypes, defaultSelection, this);
        if(isEditMode()) {
            adapter.setLightColorScheme(false);
        }

        if(isEditMode()) {
            list.setDivider(new ColorDrawable(getResources().getColor(R.color.white_with_30)));
            list.setDividerHeight(1);
        }
        list.setAdapter(adapter);
    }

    @Override
    public void onSelectionChanged() {
        if(rooms != null) {
            selection = rooms.get(adapter.getSelectedItem());
            haloController.setRoomType(selection);
        }
    }
}

File: app/src/main/java/arcus/app/pairing/device/customization/halo/room/HaloRoomFragment.kt
/*
 *  Copyright 2019 Arcus Project.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package arcus.app.pairing.device.customization.halo.room

import android.content.Context
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.core.view.ViewCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import arcus.app.R
import android.widget.Button
import arcus.app.common.view.ScleraTextView
import arcus.app.pairing.device.customization.CustomizationNavigationDelegate
import arcus.app.pairing.device.customization.favorite.FavoritesFragment
import arcus.app.common.fragment.TitledFragment
import arcus.presentation.pairing.device.customization.CustomizationStep
import arcus.presentation.pairing.device.customization.CustomizationType
import arcus.presentation.pairing.device.customization.halo.room.HaloRoom
import arcus.presentation.pairing.device.customization.halo.room.HaloRoomContract
import arcus.presentation.pairing.device.customization.halo.room.HaloRoomPresenterImpl
import org.slf4j.LoggerFactory

class HaloRoomFragment : Fragment(),
    TitledFragment,
    HaloRoomContract.View,
    HaloRoomAdapter.HaloRoomAdapterCallback {

    private lateinit var pairingDeviceAddress: String
    private lateinit var customizationStep: CustomizationStep
    private lateinit var stepTitle: ScleraTextView
    private lateinit var stepDescription: ScleraTextView
    private lateinit var nextButton: Button
    private lateinit var cancelButton: Button
    private lateinit var adapterCallback : HaloRoomAdapter.HaloRoomAdapterCallback
    private lateinit var mCallback: CustomizationNavigationDelegate
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter : HaloRoomAdapter
    private lateinit var haloRooms : List<HaloRoom>
    private val presenter : HaloRoomContract.Presenter =
        HaloRoomPresenterImpl()
    private var cancelPresent: Boolean = false
    private var nextButtonText: Int = R.string.pairing_next
    private var selectedRoom : HaloRoom? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        arguments?.let{ bundle ->
            pairingDeviceAddress = bundle.getString(FavoritesFragment.ARG_PAIRING_DEVICE_ADDRESS)!!
            customizationStep = bundle.getParcelable(FavoritesFragment.ARG_CUSTOMIZATION_STEP)!!
            cancelPresent = bundle.getBoolean(FavoritesFragment.ARG_CANCEL_PRESENT)
            nextButtonText = bundle.getInt(FavoritesFragment.ARG_NEXT_BUTTON_TEXT)
        }
    }
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_halo_room, container, false)

        stepTitle = view.findViewById(R.id.step_title)
        stepDescription = view.findViewById(R.id.step_description)

        recyclerView = view.findViewById(R.id.step_modes_list)
        recyclerView.layoutManager = LinearLayoutManager(activity)
        recyclerView.setHasFixedSize(false)
        ViewCompat.setNestedScrollingEnabled(recyclerView, false)

        nextButton = view.findViewById(R.id.next_btn)
        nextButton.setOnClickListener {
            mCallback.navigateForwardAndComplete(CustomizationType.ROOM)
            selectedRoom?.let {
                presenter.setRoom(it)
            }
        }

        cancelButton = view.findViewById(R.id.cancel_btn)
        if (cancelPresent) {
            cancelButton.visibility = View.VISIBLE
            cancelButton.setOnClickListener {
                mCallback.cancelCustomization()
            }
        } else {
            cancelButton.visibility = View.GONE
        }

        return view
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)

        adapterCallback = this
        activity?.let {
            try {
                mCallback = it as CustomizationNavigationDelegate
            } catch (exception: ClassCastException){
                logger.debug(it.toString() +
                        " must implement CustomizationNavigationDelegate: \n" +
                        exception.message)
                throw (exception)
            }
        }
    }

    override fun onResume() {
        super.onResume()

        presenter.setView(this)
        presenter.loadFromPairingDevice(pairingDeviceAddress)

        stepTitle.text = customizationStep.title ?: getString(R.string.halo_title_default)
        stepDescription.text = if(customizationStep.description.isNotEmpty()) {
            customizationStep.description.joinToString("\n")
        } else {
            resources.getString(R.string.halo_description_default)
        }
    }

    override fun getTitle(): String = customizationStep.header ?: getString(R.string.halo_header_default)

    override fun onRoomsLoaded(rooms: List<HaloRoom>) {
        haloRooms = rooms
        val firstSelection = haloRooms.indexOf(presenter.getCurrentRoomSelection())

        adapter = HaloRoomAdapter(
            firstSelection,
            haloRooms,
            adapterCallback
        )
        recyclerView.adapter = adapter
    }

    override fun onSelectionChanged() {
        selectedRoom = haloRooms[adapter.lastSelection]
    }

    override fun onError(throwable: Throwable) {
        logger.error("Halo Room Selection Customization", "Error Received", throwable)
    }

    companion object {
        private const val ARG_PAIRING_DEVICE_ADDRESS = "ARG_PAIRING_DEVICE_ADDRESS"
        private const val ARG_CUSTOMIZATION_STEP = "ARG_CUSTOMIZATION_STEP"
        private const val ARG_CANCEL_PRESENT = "ARG_CANCEL_PRESENT"
        private const val ARG_NEXT_BUTTON_TEXT = "ARG_NEXT_BUTTON_TEXT"

        @JvmStatic
        private val logger = LoggerFactory.getLogger(HaloRoomFragment::class.java)

        @JvmStatic
        fun newInstance(pairingDeviceAddress: String,
                        step: CustomizationStep,
                        cancelPresent: Boolean = false,
                        nextButtonText: Int = R.string.pairing_next
        ) : HaloRoomFragment {
            val fragment =
                HaloRoomFragment()
            with (fragment) {
                val args = Bundle()
                args.putString(ARG_PAIRING_DEVICE_ADDRESS, pairingDeviceAddress)
                args.putParcelable(ARG_CUSTOMIZATION_STEP, step)
                args.putBoolean(ARG_CANCEL_PRESENT, cancelPresent)
                args.putInt(ARG_NEXT_BUTTON_TEXT, nextButtonText)
                arguments = args
                retainInstance = true
            }
            return fragment
        }
    }
}
